dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.server.master.runner;
+
+import org.apache.dolphinscheduler.dao.entity.WorkerGroup;
+import org.apache.dolphinscheduler.dao.utils.WorkerGroupUtils;
+import org.apache.dolphinscheduler.server.master.cluster.WorkerGroupChangeNotifier;
+import org.apache.dolphinscheduler.server.master.engine.task.client.ITaskExecutorClient;
+import org.apache.dolphinscheduler.server.master.engine.task.runnable.ITaskExecutionRunnable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.PostConstruct;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * WorkerGroupTaskDispatcherManager is responsible for managing the task dispatching for worker groups.
+ * It maintains a mapping of worker groups to their task dispatchers and priority delay queues,
+ * and supports adding tasks, starting and stopping worker groups, as well as cleaning up resources upon shutdown.
+ */
+@Component
+@Slf4j
+public class WorkerGroupTaskDispatcherManager implements AutoCloseable, WorkerGroupChangeNotifier.WorkerGroupListener {
+
+    @Autowired
+    private ITaskExecutorClient taskExecutorClient;
+
+    @Getter
+    private final ConcurrentHashMap<String, WorkerGroupTaskDispatcher> dispatchWorkerMap;
+
+    public WorkerGroupTaskDispatcherManager() {
+        dispatchWorkerMap = new ConcurrentHashMap<>();
+    }
+
+    @PostConstruct
+    public void init() {
+        this.addWorkerGroup(WorkerGroupUtils.getDefaultWorkerGroup());
+    }
+
+    /**
+     * Adds a task to the specified worker group queue and starts or wakes up the corresponding processing loop.
+     *
+     * @param workerGroup the identifier for the worker group, used to distinguish different task queues
+     * @param taskExecutionRunnable an instance of ITaskExecutionRunnable representing the task to be executed
+     * @param delayTimeMills the delay time before the task is executed, in milliseconds
+     * @return true if the task is successfully added to the queue, false workerGroupTaskDispatcher not found
+     */
+    public synchronized boolean addTaskToWorkerGroup(String workerGroup, ITaskExecutionRunnable taskExecutionRunnable,
+                                                     long delayTimeMills) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.addTaskToWorkerGroupQueue(taskExecutionRunnable, delayTimeMills);
+            return true;
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found, will set task {} fail",
+                    workerGroup, taskExecutionRunnable.getTaskInstance().getId());
+        }
+        return false;
```

Comment:
Directly use computeIfAbsent then you can remove the synchronized

Reply:
To ensure that the entry point for thread creation in WorkerGroupTaskDispatcher is unique and only triggered through addWorkerGroup, special handling is required for the default worker group since it does not trigger an adding event. Moving the logic to check for the existence of a queue when adding a task could complicate the timing of additions and deletions.